src/cargo/ops/registry/publish.rs

```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
 
-    opts.gctx
-        .shell()
-        .status("Uploading", pkg.package_id().to_string())?;
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+    while !to_confirm.is_empty() {
+        if !ready.is_empty() {
+            opts.gctx.shell().status(
+                "Uploading",
+                ready.iter().map(PackageId::to_string).sorted().join(", "),
+            )?;
+        }
+
+        for pkg_id in &ready {
+            let (pkg, (_features, tarball)) = &pkg_dep_graph.packages[pkg_id];
+
+            if !opts.dry_run {
+                let ver = pkg.version().to_string();
+
+                tarball.file().seek(SeekFrom::Start(0))?;
+                let hash = cargo_util::Sha256::new()
+                    .update_file(tarball.file())?
+                    .finish_hex();
+                let operation = Operation::Publish {
+                    name: pkg.name().as_str(),
+                    vers: &ver,
+                    cksum: &hash,
+                };
+                registry.set_token(Some(auth::auth_token(
+                    &opts.gctx,
+                    &source_ids.original,
+                    None,
+                    operation,
+                    vec![],
+                    false,
+                )?));
+            }
+
+            transmit(
+                opts.gctx,
+                ws,
+                pkg,
+                tarball.file(),
+                &mut registry,
+                source_ids.original,
+                opts.dry_run,
+            )?;
+
+            if !opts.dry_run {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
+            }
+        }
+
+        let finished = if opts.dry_run {
+            ready.clone()
         } else {
-            DEFAULT_TIMEOUT
+            const DEFAULT_TIMEOUT: u64 = 60;
+            let timeout = if opts.gctx.cli_unstable().publish_timeout {
+                let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
+                timeout.unwrap_or(DEFAULT_TIMEOUT)
+            } else {
+                DEFAULT_TIMEOUT
+            };
+            if 0 < timeout {
+                let timeout = Duration::from_secs(timeout);
+                wait_for_publish(opts.gctx, source_ids.original, &to_confirm, timeout)?
```

Comment:
Publish and waiting order is non-deterministic.  `ready` is created from a `HashMap` and `to_confirm` is a `HashSet`.

Internally, this makes testing a mess (likely the tests will break if `.unordered()` isn't used).

For users, its generally nice to put things in a deterministic order.

For `wait_for_publish`, we should probably wait in the order of upload to reduce the risk of hitting the timeout unnecessarily.

Reply:
I'll sort both `ready` and `to_confirm` so they'll be in the same order. In case it wasn't clear, btw, `wait_for_publish` polls all the packages on each run through the loop.